### ISSUE TITLE
Terminal-Icons is very slow on network shares

### ISSUE THREAD

**Issue by marrsalles** [devblackops/Terminal-Icons#48]
Terminal-Icons is very slow on network shares

## Expected Behavior
Listing a network share with Terminal-Icons enabled or not should result in similar running times.

## Current Behavior
Listing a network share with Terminal-Icons enabled results in a running time that is significantly slower than the same listing executed with Terminal-Icons disabled. As an example, listing a directory with 15 entries on a network share with 20ms latency with Terminal-Icons enabled takes two seconds to complete, in contrast with a 100 milliseconds duration for the same listing with Terminal-Icons disabled. In effect, for this particular example, Terminal-Icons is 20x slower than the default behavior.

## Steps to Reproduce
1. Elect a network share with some latency and a sufficient number of entries for running the tests;
1. List the network share with Terminal-Icons disabled;
2. List the network share with Terminal-Icons enabled;
3. Compare the time it takes for the two commands to complete.

## Your Environment
Terminal-Icons 0.5.2
Windows 10 20H1
Windows Powershell 5.1

**Comment by devblackops** [comment]
Thanks for the issue @marrsalles.

I'm not able to reproduce the extreme latency you're seeing. Here is what I did:

1. Create Azure storage account with SMB file share in `eastus` (which has about 75ms latency for me).
2. Mounted file share as `z:` drive
3. Created 100 directories
4. Ran `measure-command {ls}`

With SMB info cached:
Terminal-Icons: `160ms`
No Terminal-Icons: `4ms`

I'm not aware of anything in Terminal-Icons that would require a roundtrip to the SMB share. I do call `PSIsContainer` on each `IO.FileSystemInfo` object that `Get-ChildItem` returns. Perhaps accessing that property could generate a round trip operation if information isn't in the SMB cache. I'll play around a bit more.

**Comment by DavidZidar** [comment]
I just noticed I have this issue as well. I can literally see it list the files one at a time, about 10-20 or so a second.

I don't see any noticeable delay when listing directories but files get much slower on a network drive. If I re-run the ls command it's still slow but noticeably faster the second time around. I'm guessing metadata is cached.

Calling it while wrapped in `Measure-Command` makes it go as fast as without Terminal-Icons, so that's not helpful.

**Comment by mattcargile** [comment]
I've been playing with a bit. I've concluded some of the slowness is attributed to `$_.PSIsContainer` within a custom `ps1xml` file that had an `if` block within it in like, "`if (-not $_.PSIsContainer)`".

I did try `ls | fw` it was faster. I removed the `if` check above and it worked faster on `pwsh.exe --version` , `7.2.1`. I assume failures within a `[scriptblock]` within the `ps1xml` file are converted to empty strings so I didn't need that `if` check. Maybe there is some caching involved too, like you all said, because mine sped up the more I played with it. It could have also been my VPN and network. 

After typing this comment up, I'm not sure all the way if I did something that helped.

**Comment by JensGJ** [comment]
I have the same issue. Here is a couple of findings: 

The file size of the individual file is clearly playing a role in the processing time for terminal icons. In my sample directory a single file of 16 MB takes +30 seconds to list - whereas listing 63 files less than 1 MB each takes less than 28 seconds. 
Just by looking at the output it is clear that the larger the file is the longer it takes to render with terminal icons activated. 
Without terminal icons the entire directory is listed immediately. 

The slow performance is present both in Powershell 7 and Windows Powershell 5 (7.2.6 and 5.1 to be specific).

The problem might be related to connections on VPN - I recall having had similar issues before, but haven't tested on the company network again. 

The FAQ on Oh My Posh mentions that an exclusion on Windows Defender needs to be configured. This is not the case in my setup - but Oh My Posh runs fine if Terminal Icons is not enabled. 

Any help will be appreciated.

**Comment by JensGJ** [comment]
Tested again without VPN. Still noticeably slower with terminal icons activated than without, but way faster than on my home connection with VPN (approximately 7 seconds for listing the same directory that took +1 minute from home).

**Comment by mattcargile** [comment]
Yeah I just end up living with it and I sometimes just run `ls | fw` which can be faster. I also use `PowerShellHumanizer` and even more customized formatter which slows me down a bit more. Not sure much can be done to speed up the formatter. Most of the time it works well enough and the colors and icons are worth it.

**Comment by DavidZidar** [comment]
In case anyone is interested in only the colors like I was here's my configuration inspired by this project but using the experimental PowerShell 7.2 feature [PSANSIRenderingFileInfo](https://learn.microsoft.com/en-us/powershell/scripting/learn/experimental-features?view=powershell-7.2#psansirenderingfileinfo). No performance issues, works great imo, no icons though.
https://gist.github.com/DavidZidar/1cfef361ff659aef68cef25ebd547fa9

**Comment by Tiberriver256** [comment]
Potentially faster:

`[system.io.file]::GetAttributes($_.FullName) -eq [System.IO.FileAttributes]::Directory`

The downside is it only works on file systems. It'll throw an exception if you try and run it on something returned from `dir HKCU:` for example.

**Comment by JensGJ** [comment]
For testing purpose I've used the following: 

`$timer = [Diagnostics.Stopwatch]::StartNew(); ls ; $timer.stop(); $timer`

(since measure-command as mentioned earlier cannot be used because the slow performance is only present when the ls output is displayed in the console).

**Comment by mattcargile** [comment]
> I'm not sure if it's related with what you're all describing but I have a local folder with lots of LARGE .zip files and using this module sometimes triggers my antivirus to check those files when I `ls` (if I check the processes I can see kaspersky using a lot of disk resources) and the files are listed to me one by one. If I try with `ls | fw` or I unload the module and try `ls` again, then the files are listed instantly.
> 
> edit: I also made a module based on @DavidZidar idea, using the current colors from Terminal-Icons (except for default folder), considering the limitations of PSANSIRenderingFileInfo (like you can't set colors for specific files or folder names etc). Here https://gist.github.com/rafma0/ba70b41e3a100b88307703408dd7a795

Interesting. I didn't think about the AV. I currently use _Sentinel One_ which doesn't appear to be as bad as _Windows Defender_ ( e.g. `MsMpEng.exe` ). I also used to use a VPN that had to go out of state. I'm also using the module `PowerShellHumanizer`. Lately, I deal with the slowness, but it really hasn't been as bad lately. I love the colors and "pretty" file sizes too much.
